### PR TITLE
Always use `auth_servers` when proxying an auth connection

### PR DESCRIPTION
**Pull request by espadolini**
As a followup to #12862, this PR makes it so that every instance of dialing an auth through a proxy (not just the reverse tunnel's `@remote-auth-server`) will select one random entry from the proxy's own `auth_servers` configuration rather than checking the auth server heartbeats in the backend. This is preferrable as the backend info could be out of date (10 minutes for expiration plus backend expiration and cache propagation delays), and doesn't allow for any flexibility in choosing a specific load balancer or running health checks.

note: backporting this requires backporting #12862 at the same time.

### CODE REVIEW

File: lib/reversetunnel/localsite.go
Comment (codingllama) [inline lib/reversetunnel/localsite.go:179]: Are there any implications in dropping this loop that we should be aware of?

File: lib/srv/alpnproxy/auth/auth_proxy.go
Comment (codingllama) [inline lib/srv/alpnproxy/auth/auth_proxy.go:120]: Ditto here, any implications in dropping this loop?

File: lib/srv/alpnproxy/auth/auth_proxy_test.go
Comment (codingllama) [inline lib/srv/alpnproxy/auth/auth_proxy_test.go:29]: nit: can we get some inline comments so the meaning is clearer?

```suggestion
	s := NewAuthProxyDialerService(nil /* foo */, "" /* bar */, nil  /* authServers */)
```

File: lib/utils/utils.go
Comment (codingllama) [inline lib/utils/utils.go:564]:
```suggestion
	switch len(slice) {
	case 0:
		return ""
	case 1:
		return slice[0]
	default:
		return slice[rand.Intn(len(slice))]
	}
```

**Review by codingllama — APPROVED** [review]
LGTM.

This isn't a part of Teleport that I usually inhabit, so something might escape me, but I like that the change is both simpler and makes a clearly random choice.

File: lib/srv/alpnproxy/auth/auth_proxy_test.go
Comment (codingllama) [reply lib/srv/alpnproxy/auth/auth_proxy_test.go:45]: My 2c: I'm not sure Cleanups are built to handle the eventual t.Fatal, so I wouldn't do this.

Comment (espadolini) [reply lib/srv/alpnproxy/auth/auth_proxy_test.go:45]: I'm pretty sure that `t.FailNow` is handled correctly by `t.Cleanup` - I'm not sure if the rest of that particular cleanup callback is executed or if it's Goexited out, but the other callbacks should still run afterwards, even.

Comment (codingllama) [reply lib/srv/alpnproxy/auth/auth_proxy_test.go:45]: IMO, if the error is relevant, then it belongs in the body of the test. It doesn't look this one is.

I wouldn't encourage reproducing this pattern, point being that neither of us seem certain of what happens with other cleanups.

File: lib/reversetunnel/localsite.go
Comment (espadolini) [reply lib/reversetunnel/localsite.go:179]: All our current clients that dial `@remote-auth-addr` through a reverse tunnel connection have builtin retries, so even in the case of multiple `auth_servers` specified in the configuration file (which is very much discouraged) this is going to be fine.

File: lib/srv/alpnproxy/auth/auth_proxy.go
Comment (espadolini) [reply lib/srv/alpnproxy/auth/auth_proxy.go:120]: This change (and the one in `proxySubsys`) are more controversial; we discussed this in the scaling meeting and the consensus was that client (this will affect `tsh` and Connect) connection failures with a properly configured auth load balancer (the HA setup we recommend and support) will be few and far between, and any retry logic should be implemented in the client, if it turns out to be necessary. Another point in favor of only triggering one connection attempt per client connection is to prevent network amplification impact from a malicious or misconfigured client - or at least reduce it, as `net.Dial` on a hostname might attempt to connect to all the resolved addressed in sequence, but we don't specialcase that anywhere in the codebase.

**Review by LKozlowski — APPROVED** [review]
LGTM.

I'm not very familiar with this part of code base, but just one thing came to my mind. 

In the previous version we were checking authorization servers for connection one by one. Now, we are picking up one server randomly and if it fails then we throws and error. There's no fallback or checking the other servers from the list. Isn't it going to affect the availability?

File: lib/srv/alpnproxy/auth/auth_proxy_test.go
Comment (espadolini) [reply lib/srv/alpnproxy/auth/auth_proxy_test.go:45]: `require.Foo` is used in `t.Cleanup` throughout our codebase and it's a common recommended pattern in a bunch of guides. The interaction between `t.Cleanup` and `t.FailNow` is indeed as I was expecting: the specific cleanup function is interrupted by `runtime.Goexit` but the remaining cleanup functions are still executed (`(*testing.common).runCleanup` will keep calling itself from a deferred function while `c.cleanups` is nonempty).

It's totally fine to ignore errors on close in the `Cleanup` in this comment chain (and I might leave it to be ignored as is), but it's not a pattern to be avoided in general.

**Comment by espadolini** [comment]
@LKozlowski 
> There's no fallback or checking the other servers from the list. Isn't it going to affect the availability?

See https://github.com/gravitational/teleport/pull/13310#discussion_r894321569, the tl;dr is that it should be quite rare and only for `tsh` - if it turns out to be a problem, any retry logic should be in `tsh` itself rather than hidden away in the proxy. We never noticed problems because the usual setup in kubernetes will notify the proxy of a failure to connect to auth really quickly, and the retry logic masked the design flaw of using the backend info over the explicit user configuration.

**Comment by espadolini** [comment]
@codingllama
> makes a clearly random choice

I specialcased length 1 in the utility function specifically because in almost every configuration (and in all supported ones) there will only be a single address in `auth_servers`; the random choice is just there to still guarantee the correct load balancing behavior for old and/or weird setups in which proxies have a hardcoded list of more than one auth, but we don't really expect it to happen.

For a DNS load balancer setup, `net.Dial` itself will connect to the resolved DNS entries (if there's more than one) in order, but that's because the entries from the DNS server are supposed to be in random order.

**Comment by LKozlowski** [comment]
> 



> @LKozlowski
> 
> > There's no fallback or checking the other servers from the list. Isn't it going to affect the availability?
> 
> See [#13310 (comment)](https://github.com/gravitational/teleport/pull/13310#discussion_r894321569), the tl;dr is that it should be quite rare and only for `tsh` - if it turns out to be a problem, any retry logic should be in `tsh` itself rather than hidden away in the proxy. We never noticed problems because the usual setup in kubernetes will notify the proxy of a failure to connect to auth really quickly, and the retry logic masked the design flaw of using the backend info over the explicit user configuration.

Thanks for the explanation! Makes sense now. I must have written my comment when you were writing yours as I didn't notice it before sending my feedback 😄

File: lib/srv/alpnproxy/auth/auth_proxy.go
Comment (espadolini) [reply lib/srv/alpnproxy/auth/auth_proxy.go:111]: I'm just going to `Wrap` the existing error as is.

File: lib/srv/alpnproxy/auth/auth_proxy_test.go
Comment (espadolini) [reply lib/srv/alpnproxy/auth/auth_proxy_test.go:45]: I'm adding the `require.NoError`.

Comment (espadolini) [reply lib/srv/alpnproxy/auth/auth_proxy_test.go]: Of course, by "a bunch" I clearly meant "one": https://docs.docker.com/language/golang/run-tests/ 😬 

There's also handful of examples around the internet that use `stretchr/testify/require` in `t.Cleanup`, however.

Comment (espadolini) [reply lib/srv/alpnproxy/auth/auth_proxy_test.go]: Technically if your `require.Foo` is the last thing in the function, you can swap it out for an `assert.Foo` instead.

**Review by rosstimothy — APPROVED** [review]

**Comment by espadolini** [comment]
Holding off on the v7 backport.

**Pull request merged**
